File: research/academic_tools.py
"""Academic search tools using Semantic Scholar and ArXiv HTTP APIs."""

import asyncio
import logging
import urllib.parse

import httpx
from langchain_core.tools import tool

from research.schemas import PaperRef

logger = logging.getLogger(__name__)

SEMANTIC_SCHOLAR_BASE = "https://api.semanticscholar.org/graph/v1"
ARXIV_BASE = "https://export.arxiv.org/api"

_FIELDS = "paperId,title,abstract,authors,year,citationCount,externalIds"


def _to_paper_ref(item: dict) -> PaperRef:
    """Convert a Semantic Scholar API response item to PaperRef."""
    doi = ""
    ext = item.get("externalIds") or {}
    if isinstance(ext, dict):
        doi = ext.get("DOI", "")

    return PaperRef(
        id=item.get("paperId", ""),
        title=item.get("title", ""),
        abstract=item.get("abstract", "") or "",
        authors=[a.get("name", "") for a in (item.get("authors") or [])],
        year=item.get("year") or 0,
        citation_count=item.get("citationCount") or 0,
        doi=doi,
    )


@tool
def search_semantic_scholar(query: str, limit: int = 10) -> list[dict]:
    """Search Semantic Scholar for academic papers.

    Args:
        query: Search query string
        limit: Maximum number of results (default 10, max 100)

    Returns:
        List of paper dicts with id, title, abstract, authors, year, citation_count
    """
    params = {
        "query": query,
        "limit": min(limit, 100),
        "fields": _FIELDS,
    }
    try:
        with httpx.Client(timeout=30) as client:
            resp = client.get(f"{SEMANTIC_SCHOLAR_BASE}/paper/search", params=params)
            resp.raise_for_status()
            data = resp.json()
            papers = [_to_paper_ref(item).model_dump() for item in data.get("data", [])]
            logger.info("semantic_scholar: %d results for '%s'", len(papers), query)
            return papers
    except httpx.HTTPError as exc:
        logger.warning("semantic_scholar search failed: %s", exc)
        return []


@tool
def search_arxiv(query: str, limit: int = 10) -> list[dict]:
    """Search ArXiv for academic preprints.

    Args:
        query: Search query (supports ArXiv query syntax)
        limit: Maximum number of results

    Returns:
        List of paper dicts with id, title, abstract, authors, year
    """
    import xml.etree.ElementTree as ET

    params = {
        "search_query": f"all:{urllib.parse.quote(query)}",
        "max_results": min(limit, 100),
        "sortBy": "relevance",
    }
    try:
        with httpx.Client(timeout=30) as client:
            resp = client.get(f"{ARXIV_BASE}/query", params=params)
            resp.raise_for_status()

        ns = {"atom": "http://www.w3.org/2005/Atom"}
        root = ET.fromstring(resp.text)
        papers = []
        for entry in root.findall("atom:entry", ns):
            title = (entry.findtext("atom:title", "", ns) or "").strip()
            abstract = (entry.findtext("atom:summary", "", ns) or "").strip()
            arxiv_id = (entry.findtext("atom:id", "", ns) or "").split("/abs/")[-1]
            authors = [
                a.findtext("atom:name", "", ns) or ""
                for a in entry.findall("atom:author", ns)
            ]
            published = entry.findtext("atom:published", "", ns) or ""
            year = int(published[:4]) if published else 0

            papers.append(PaperRef(
                id=f"arxiv:{arxiv_id}",
                title=title,
                abstract=abstract,
                authors=authors,
                year=year,
            ).model_dump())

        logger.info("arxiv: %d results for '%s'", len(papers), query)
        return papers
    except Exception as exc:
        logger.warning("arxiv search failed: %s", exc)
        return []


ACADEMIC_TOOLS = [search_semantic_scholar, search_arxiv]

File: src/research/academicTools.ts
import { getConfig } from "../config.js";
import { withApiRetry } from "../retry.js";
import type { Paper } from "../graph/state.js";
import { getLogger } from "../observability.js";

const log = getLogger("research:academic");

const SS_BASE = "https://api.semanticscholar.org/graph/v1";
const ARXIV_BASE = "https://export.arxiv.org/api/query";

interface SsRecord {
  paperId: string;
  title: string;
  abstract?: string;
  authors?: Array<{ name: string }>;
  year?: number;
  url?: string;
}

export async function searchSemanticScholar(
  query: string,
  limit = 10,
): Promise<Paper[]> {
  const cfg = getConfig();
  const headers: Record<string, string> = {
    "User-Agent": "VFS-Research/0.1",
  };
  if (cfg.semanticScholarKey) {
    headers["x-api-key"] = cfg.semanticScholarKey;
  }

  const url = new URL(`${SS_BASE}/paper/search`);
  url.searchParams.set("query", query);
  url.searchParams.set("limit", String(limit));
  url.searchParams.set("fields", "paperId,title,abstract,authors,year,url");

  return withApiRetry(async () => {
    const res = await fetch(url.toString(), { headers });
    if (!res.ok) throw new Error(`Semantic Scholar error: ${res.status}`);
    const data = (await res.json()) as { data?: SsRecord[] };
    return (data.data ?? []).map(
      (p): Paper => ({
        paperId: p.paperId,
        title: p.title,
        abstract: p.abstract,
        authors: (p.authors ?? []).map((a) => a.name),
        year: p.year,
        url: p.url,
      }),
    );
  });
}

export async function searchArXiv(query: string, maxResults = 10): Promise<Paper[]> {
  const url = new URL(ARXIV_BASE);
  url.searchParams.set("search_query", `all:${encodeURIComponent(query)}`);
  url.searchParams.set("max_results", String(maxResults));

  return withApiRetry(async () => {
    const res = await fetch(url.toString());
    if (!res.ok) throw new Error(`ArXiv error: ${res.status}`);
    const xml = await res.text();
    return parseArXivXml(xml);
  });
}

function parseArXivXml(xml: string): Paper[] {
  const papers: Paper[] = [];
  const entries = xml.match(/<entry>([\s\S]*?)<\/entry>/g) ?? [];

  for (const entry of entries) {
    const id = entry.match(/<id>(.*?)<\/id>/)?.[1] ?? "";
    const title = entry.match(/<title>([\s\S]*?)<\/title>/)?.[1]?.trim() ?? "";
    const abstract = entry.match(/<summary>([\s\S]*?)<\/summary>/)?.[1]?.trim();
    const yearMatch = entry.match(/<published>(\d{4})/);
    const year = yearMatch ? parseInt(yearMatch[1]!, 10) : undefined;

    const authorMatches = [...entry.matchAll(/<name>(.*?)<\/name>/g)];
    const authors = authorMatches.map((m) => m[1] ?? "");

    if (id && title) {
      papers.push({ paperId: id, title, abstract, authors, year, url: id });
    }
  }

  log.debug({ count: papers.length }, "ArXiv results parsed");
  return papers;
}
